_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,181 @@
+
+We learned two things about the ViewModel the hard way:
+
+* The ViewModel should expose states for the View, rather than just events. For example, if we need to display the name and the email address of a `User`, rather than creating two streams for this, we create a `DisplayableUser` object that encapsulates the two fields. The stream will emit every time the display name or the email change. Like this, we ensure that our View always displays the current state of the `User`.
+
+* We should make sure that every action of the user goes through the ViewModel and that any possible logic of the View is moved in the ViewModel.
+
+We shared more details about these two topics in a blog post about <a href="https://upday.github.io/mvvm_rx_common_mistakes">common mistakes in MVVM + RxJava</a>.  
+
+### View
+
+The View is the actual user interface in the app. It can be an `Activity`, a `Fragment` or any custom Android `View`. For `Activities` and `Fragment`s, we are binding and unbinding from the event sources on `onResume()` and `onPause()`.
+
+{% highlight java %}
+    private final CompositeSubscription mSubscription = new CompositeSubscription();
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        mSubscription.add(mViewModel.getSomeData()
+                         .observeOn(AndroidSchedulers.mainThread())
+                         .subscribe(this::updateView,
+                                    this::handleError));
+    }
+
+    @Override
+    public void onPause() {
+        mSubscription.clear();
+        super.onPause();
+    }
+{% endhighlight %}
+
+In case the MVVM View is a custom Android `View`, the binding is done in the constructor. To ensure that the subscription is not preserved, leading to possible memory leaks, the unbinding happens in `onDetachedFromWindow`.
+
+{% highlight java %}
+    private final CompositeSubscription mSubscription = new CompositeSubscription();
+
+    public MyView(Context context, MyViewModel viewModel) {
+        ...
+        mSubscription.add(mViewModel.getSomeData()
+                         .observeOn(AndroidSchedulers.mainThread())
+                         .subscribe(this::updateView,
+                                    this::handleError));
+    }
+
+    @Override
+    public void onDetachedFromWindow() {
+        mSubscription.clear();
+        super.onDetachedFromWindow();
+    }
+}
+{% endhighlight %}
+
+## Testability Of The Model-View-ViewModel Classes
+
+One of the main reasons why we love the Model-View-ViewModel pattern is its high degree of testability.
+
+### DataModel
+
+The inversion of control pattern, heavily applied in our code, and the lack of any Android classes, make the DataModel easy to unit test.
+
+### ViewModel
+
+The ViewModel exposes streams of data to whoever needs to consume it. We see the Views and the unit tests as two different types of data consumers.
```

Comment:
If you want to word more like you are explaining the concept, say -
> We can see the ...

If you want to tell the reader how we do it at upday, your wording is better.